compiler/rustc_resolve/src/ident.rs

```diff
@@ -598,7 +600,36 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
                         result
                     }
                     Scope::BuiltinTypes => match this.builtin_types_bindings.get(&ident.name) {
-                        Some(binding) => Ok((*binding, Flags::empty())),
+                        Some(binding) => {
+                            match ident.name {
+                                sym::f16
+                                    if !this.tcx.features().f16
+                                        && !ident.span.allows_unstable(sym::f16) =>
+                                {
+                                    feature_err(
+                                        this.tcx.sess,
+                                        sym::f16,
+                                        ident.span,
+                                        "the type `f16` is unstable",
+                                    )
+                                    .emit();
+                                }
+                                sym::f128
+                                    if !this.tcx.features().f128
+                                        && !ident.span.allows_unstable(sym::f128) =>
+                                {
+                                    feature_err(
+                                        this.tcx.sess,
+                                        sym::f128,
+                                        ident.span,
+                                        "the type `f128` is unstable",
+                                    )
+                                    .emit();
+                                }
```

Comment:
This is where the error is coming from but I am not sure why. In `tests/ui/resolve/conflicting-primitive-names.rs` it seems like the same-name types all get resolved before `Scope::BuiltinTypes` even gets checked. Not sure what exactly makes the `half` crate act differently than the test

Reply:
Oh you know what? Try checking if `force` is true, and only issuing a feature error then. Otherwise, return `Err(Determinacy::Undetermined)`.

Reply:
cc @petrochenkov who also may know better.

Reply:
Hm, returning an `Err(Determinacy::Undetermined)` from that block (in place of the `Ok((*binding, Flags::empty()))`) hits an ice at `compiler/rustc_resolve/src/lib.rs:2064`, need to dig a bit more.

What does `force` do here? I'm not sure how to replicate this in a unit test

Reply:
Well changing https://github.com/rust-lang/rust/blob/b8f45a3189824a85750537d623820b35d9180399/compiler/rustc_resolve/src/lib.rs#L2064 to `None` makes it compile with no ICEs and no FP on the gate lol. I need to trace back why that branch is expected to never hit

Reply:
Actually, `Err(Determinacy::Determined)` "seems" right here. Unfortunately, literally none of this is documented as far as I can tell.